hostap: fix VHT channel center segment0

gangli02 opened the pull request:
hostap: fix VHT channel center segment0

1) Unless ACS is being used, both "channel" and
    "vht_oper_centr_freq_seg0_idx" parameters must be set.
    Fixed "channel center segment 0" not being set in VHT Operation IE.
2) Set HT capabilities and VHT capabilities via the wifi ap config command.
3) Set AP bandwidth to wifi driver via the wifi_mgmt_api->ap_config_params.

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -2082,6 +2082,78 @@ int hapd_state(const struct device *dev, int *state)
+		center_freq_seg0_idx = params->channel;
+		break;
+	case WIFI_FREQ_BANDWIDTH_40MHZ:
+		if (!hostapd_cli_cmd_v("set ht_capab [HT40+][SHORT-GI-40]")) {
```
SHORT-GI  should be set as per the underlying driver/chipset capablities

krish2718 replied:
the choice of HT40+/- should still be user configurable, no? 36-40MHz-HT40+ === 40-40MHz-HT40- == 38.

gangli02 replied:
When starting soft AP, it cannot get the driver/chipset capabilities. 
What do you think about putting these configurations in the "wifi ap config" command?
Before starting the AP, users can freely set it, similar to the hostapd.conf file.

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -2082,6 +2082,78 @@ int hapd_state(const struct device *dev, int *state)
+		return -EINVAL;
+	}
+
+	if (params->bandwidth != WIFI_FREQ_BANDWIDTH_20MHZ)
+	{
+		chan_idx = params->channel;
+		for (index = 0; index < index_max; index++)
+		{
+			if ((chan_idx >= (center_freq[index] - ch_offset)) && (chan_idx <= (center_freq[index] + ch_offset)))
+			{
+					center_freq_seg0_idx = center_freq[index];
+					break;
+			}
+		}
+	}
```
can't we just do below?

```
	if (params->bandwidth != WIFI_FREQ_BANDWIDTH_20MHZ)
	{
		center_freq_seg0_idx = params->channel + ch_offset;
	}
```
We don't even need the static array of frequencies? Of course `+/-` depends on the HT40+/- user input.

gangli02 replied:
Yes, I will check it.

krish2718 reviewed (changes requested):
Overall we need a lot more use inputs to do the freq-channel conversions e.g., HT40+/-, regulatory classes etc.

gangli02 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -560,6 +560,8 @@ struct wifi_connect_req_params {
 	/** User Passwords */
 	const uint8_t *passwords[WIFI_ENT_IDENTITY_MAX_USERS];
+	/** Parameter used for frequency band */
+	enum wifi_frequency_bandwidths bandwidth;
```
Sure

MaochenWang1 commented:
FYI @nxf58150

gangli02 committed:
hostap: fix VHT channel center segment0

1) Unless ACS is being used, both "channel" and
"vht_oper_centr_freq_seg0_idx" parameters must be set.
Fixed "channel center segment 0" not being set in VHT Operation IE.
2) Set HT capabilities and VHT capabilities via the wifi ap config command.
3) Set AP bandwidth to wifi driver via the wifi_mgmt_api->ap_config_params.

Signed-off-by: Gang Li <gang.li_1@nxp.com>

gangli02 committed:
drivers: wifi: nxp: set the maximum number of stations

Add support for setting the maximum number of stations.
Add support for setting AP bandwidth.

Signed-off-by: Gang Li <gang.li_1@nxp.com>

gangli02 commented:
@krish2718  please review again, thanks!

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -655,6 +676,14 @@ enum wifi_ap_config_param {
 	WIFI_AP_CONFIG_PARAM_MAX_NUM_STA = BIT(1),
+	/** Used for AP mode configuration parameter bandwidth */
+	WIFI_AP_CONFIG_PARAM_BANDWIDTH = BIT(2),
+#if defined(CONFIG_WIFI_NM_WPA_SUPPLICANT)
```
any reason for this guard? These could be applicable to offload too?

gangli02 replied:
Yes, no need.

gangli02 replied:
> Currently this parameter is related to hostap. Should we consider wrapping it with `CONFIG_WIFI_NM_HOSTAPD_AP` macro?

@krish2718, what do you think?

krish2718 replied:
Sure, WPA supplicant based AP doesn't provide these anyway.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1821,11 +1841,15 @@ static int wifi_ap_config_args_to_params(const struct shell *sh, size_t argc, ch
 		{"max_inactivity", required_argument, 0, 'i'},
 		{"max_num_sta", required_argument, 0, 's'},
+#if defined(CONFIG_WIFI_NM_WPA_SUPPLICANT)
+		{"ht_capab", required_argument, 0, 'n'},
+		{"vht_capab", required_argument, 0, 'c'},
```
`htc` and `vhtc`  as short forms?

gangli02 replied:
Sorry, which one?  
If you mean `ht_capab` -> `htc`,  here `ht_capab` is to be consistent with the hostapd.conf file. If the user wants to set it, they can directly refer to the hostapd.conf file.
If you mean `n`->`htc`, strings shouldn't be used here, right?

krish2718 replied:
> If you mean n->htc, strings shouldn't be used here, right?

I was referring to this, if you use `getopt_long` then you can use multicharacter short options, but anyways not mandatory.

MaochenWang1 commented on modules/hostap/src/supp_api.c:
```diff
@@ -2486,6 +2585,12 @@ int supplicant_ap_enable(const struct device *dev,
 		return -1;
 	}
 
+	ret = set_ap_bandwidth(dev, params->bandwidth);
```
should this be after `k_mutex_lock(&wpa_supplicant_mutex, K_FOREVER);`?

MaochenWang1 commented on modules/hostap/src/supp_api.c:
```diff
@@ -2486,6 +2585,12 @@ int supplicant_ap_enable(const struct device *dev,
 
+	ret = set_ap_bandwidth(dev, params->bandwidth);
+	if (ret) {
+		wpa_printf(MSG_ERROR, "Failed to set ap bandwidth");
```
If other platforms don't support offload `ap_config_params`, then when it starts AP, it will fail and return here?

gangli02 replied:
Yes, will update.

MaochenWang1 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -986,6 +989,14 @@ struct wifi_ap_config_params {
 	uint32_t max_num_sta;
+	/** Parameter used for frequency band */
+	enum wifi_frequency_bandwidths bandwidth;
+#if defined(CONFIG_WIFI_NM_WPA_SUPPLICANT)
```
Should we use `CONFIG_WIFI_NM_HOSTAPD_AP` for all the `ht_capab`?
As `supplicant_ap_config_params` is wrapped by `HOSTAPD macro`.

gangli02 replied:
Yes, `CONFIG_WIFI_NM_HOSTAPD_AP` is more suitable.
And maybe it doesn't need to be wrapped with macros.
What do you think? @krish2718

gangli02 committed:
net: wifi: add "bandwidth" parameter to "wifi ap enable" command

Add "bandwidth" parameter to "wifi ap enable" command.
Add "ht_capab" and "vht_capab" parameters to "wifi ap config" command.

Signed-off-by: Gang Li <gang.li_1@nxp.com>

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -2241,53 +2311,72 @@ int hapd_config_network(struct hostapd_iface *iface,
+	k_mutex_lock(&wpa_supplicant_mutex, K_FOREVER);
+
+	iface = get_hostapd_handle(dev);
+	if (!iface) {
```
We are trying to migrate away from this style as there is a MISRA rule that recommends to use this style
`iface == NULL` for non boolean variable checks.

gangli02 replied:
OK

gangli02 commented on include/zephyr/net/wifi.h:
```diff
@@ -655,6 +676,14 @@ enum wifi_ap_config_param {
 	WIFI_AP_CONFIG_PARAM_MAX_NUM_STA = BIT(1),
+	/** Used for AP mode configuration parameter bandwidth */
+	WIFI_AP_CONFIG_PARAM_BANDWIDTH = BIT(2),
+#if defined(CONFIG_WIFI_NM_HOSTAPD_AP)
```
Yes. The ifdefs are added because these parameters are currently only supported when hostapd is enabled.

jukkar replied:
I understand, but we do not need to hide them even if we are not using them. It just makes a life a bit easier as one can  immediately see what bits are available even if they are only used by some Kconfig option.
We have had this style of coding in the past (hiding some values in enums) but have tried to migrate away from it.

gangli02 replied:
OK, I get it.

gangli02 commented on modules/hostap/src/supp_api.c:
```diff
@@ -2075,6 +2075,71 @@ int hapd_state(const struct device *dev, int *state)
+	uint8_t center_freq_seg0_idx = 0;
+	uint8_t oper_chwidth = CHANWIDTH_USE_HT;
+	uint8_t *center_freq = NULL;
+	uint8_t center_freq_40MHz[] = {38, 46, 54, 62, 102, 110, 118, 126, 134, 142, 151, 159};
+	uint8_t center_freq_80MHz[] = {42, 58, 106, 122, 138, 155};
```
Sure

gangli02 commented on modules/hostap/src/supp_api.c:
```diff
@@ -2075,6 +2075,71 @@ int hapd_state(const struct device *dev, int *state)
+	int ret = 0;
+	uint8_t center_freq_seg0_idx = 0;
+	uint8_t oper_chwidth = CHANWIDTH_USE_HT;
+	uint8_t *center_freq = NULL;
```
OK

gangli02 commented:
Hi @jukkar, please help review again
